### PR TITLE
Fix different behavior with unittest when warlus operator

### LINKED ISSUE #10743: Walrus operator causes different behavior in PyTest.
I am currently testing the following function but find that the function passes the test in the normal Python terminal but fails in a PyTest run. The walrus operator is relatively new and not many people use it. I think that there may be an inconsistency in the execution environment.

```
import numpy as np

def test_walrus_conversion():
    a = np.random.random(16)
    assert not np.array_equal(a, a := a.astype(np.uint8))
    assert np.all(a == 0)
```

### PR DESCRIPTION
**Pull request by aless10**
The `AssertionRewriter` class did not handle the assertions with the `ast.NamedExpr` (aka 'walrus operator'), that changed the value of a variable used in the assertion (an example can be seen in the related issue or in the tests added). 

This caused different behavior from unittest/python and pytest.

This PR aims to handle this case (although very remote)

_Addresses python version >= 3.8_

closes #10743

### CODE REVIEW

**Comment by aless10** [comment]
It looks like one of the checks failed but I don't know if it depends on the changes or what. What is the next action in this situation? I'd like to help in some way. Thanks

**Comment by RonnyPfannschmidt** [comment]
i restarted the job, its unrelated to your change

File: testing/test_assertrewrite.py
Comment (nicoddemus) [inline testing/test_assertrewrite.py:1377]: I noticed this as well, can't we save the first `a` value in a separate temporary variable, before evaluating the walrus operator? This would provide the same (correct) error message as in `test_assertion_walrus_no_variable_name_conflict`.

File: src/_pytest/assertion/rewrite.py
Comment (nicoddemus) [inline src/_pytest/assertion/rewrite.py:947]: This was copied from `visit_Name`, perhaps we can update the comment to mention that this handles the walrus operator instead?

**Review by nicoddemus — COMMENTED** [review]
Thanks a lot for tacking this @aless10!

Left two comments, could you take a look? Thanks!

File: testing/test_assertrewrite.py
Comment (aless10) [reply testing/test_assertrewrite.py:1377]: I tried to do so, but with my approach I did not change the value of the variable `a` but I saved the new value of `a` in another temporary variable. The problem with this is that if that variable is used later in another assert there will be an error. I think is doable, I just need to try different approach. Thanks

Comment (nicoddemus) [reply testing/test_assertrewrite.py:1377]: >  The problem with this is that if that variable is used later in another assert there will be an error

IIUC the problem, you can use `self.variable()` to obtain a new, unique variable to save the value to. This would prevent any conflicts in later asserts.

Comment (aless10) [reply testing/test_assertrewrite.py]: Thanks for the tip: I refactored using the `variable` method but I don't think this is enough to solve the problem. For example in this test case
```
def test_walrus_conversion_succeed():
    a = "Hello"
    assert a != (a := a.lower())
    assert a == 'hello'
```
if I change the name of the first "a", then I got an error because it is referencing something that has no reference (in fact "a" is defined outside the first assert statement. I tried to change the name of the new "a" in the walrus operator, but then the second assert fails because the value of "a" is still the original one. 
The only solution that came to my mind is using a map/state that lives through the asserts that enables me to retrieves the values if they have changed.
That said, this situation is very rare and the tests cases that came to my mind are far from being used in real case scenario (but this is only my opinion). I'm thinking if the juice worth the squeeze

**Comment by aless10** [comment]
I don't know why all the checks are failing. Last time they passed was at commit 3e90bf5 and it looks like the failing tests are not referenced by these changes. 
I tried to reproduce the error and actually the `tox` commands fails but if I run the tests inside a docker container, the tests pass. I'm trying to understand what's going on. Sorry for the inconvenience

File: testing/test_assertrewrite.py
Comment (aless10) [reply testing/test_assertrewrite.py]: So, after https://github.com/pytest-dev/pytest/pull/10788, all the tests pass. Right now, we have covered all of the cases that are tested in `test_assertrewrite.py::TestIssue10743` where I tried to imagine most of the corner cases. Let me know if you like my solution or if you prefer to revert and add a note for this remote case. Thanks

**Comment by aless10** [comment]
> I don't know why all the checks are failing. Last time they passed was at commit [3e90bf5](https://github.com/pytest-dev/pytest/commit/3e90bf573f4415e6b34b9187f625ef265d634aeb) and it looks like the failing tests are not referenced by these changes. I tried to reproduce the error and actually the `tox` commands fails but if I run the tests inside a docker container, the tests pass. I'm trying to understand what's going on. Sorry for the inconvenience

OK it looks like I had to align with this https://github.com/pytest-dev/pytest/pull/10788 fix. Sorry

File: changelog/10743.bugfix.rst
Comment (nicoddemus) [inline changelog/10743.bugfix.rst:1]: I think we can update this to mean that walrus operator now is properly supported by the assertion rewriting, correct?

```suggestion
The assertion rewriting mechanism now works correctly when an assertion expression contains the walrus operator.
```

Comment (aless10) [reply changelog/10743.bugfix.rst]: LGTM. I pushed the changes. Thanks

**Comment by nicoddemus** [comment]
Thanks! I will merge in a few days unless other maintainers want to take a closer look.

**Review by nicoddemus — APPROVED** [review]

**Pull request merged**
